ArmPkg/Library/ArmMmuLib/AArch64/ArmMmuLibPageTableAlloc.c

```diff
@@ -0,0 +1,121 @@
+/** @file ArmMmuLibPageTableAlloc.c
+
+Logic facilitating the pre-allocation of page table memory.
+
+Copyright (C) Microsoft Corporation. All rights reserved.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Uefi.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/ArmMmuLib.h>
+
+typedef struct {
+  VOID     *NextPool;
+  UINTN    Offset;
+  UINTN    FreePages;
+} PAGE_TABLE_POOL;
+
+PAGE_TABLE_POOL  *mPageTablePool    = NULL;
+BOOLEAN          mPageTablePoolLock = FALSE;
+
+/**
+  Initializes the reserved pool used to allocate pages for page table memory.
+
+  @param[in] PoolPages The number of pages to initialize for the reserve pool. The actual number of pages
+                       reserved will be at least 512 - the number of pages required to describe a
+                       fully split 2MB region.
+
+  @retval EFI_SUCCESS             Reserved pool initialized successfully
+  @retval EFI_INVALID_PARAMETER   PoolPages is zero
+  @retval EFI_ABORTED             Software lock is held by another call to this function
+  @retval EFI_OUT_OF_RESOURCES    Unable to allocate pages
+**/
+EFI_STATUS
+InitializePageTablePool (
+  IN  UINTN  PoolPages
+  )
+{
+  VOID        *Buffer;
+  EFI_STATUS  Status = EFI_SUCCESS;
+
+  if (PoolPages == 0) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  //
+  // Avoid an infinite loop by bailing if we are nested in another call
+  // to this function
+  //
+  if (mPageTablePoolLock) {

```

Comment:
I think I may update the reserved page table pool to be fetched via protocol and owned by Arm's CpuDxe to make it a hybrid of the x86 approach and the Arm approach to page management. This will prevent a lot of strange edge cases which could occur if more than one module include ArmMmuLib and reduce the memory footprint.

Reply:
Updated to put the reserved pool in CpuDxe. I also updated the page table memory allocation behavior to keep the number of preallocated pages above 5.